components/net/indexeddb/idb_thread.rs

```diff
@@ -0,0 +1,228 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use crate::indexeddb::engines::{
+    KvsEngine, KvsOperation, KvsTransaction, RkvEngine, StoreDescription,
+};
+
+use std::borrow::ToOwned;
+use std::cell::Cell;
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::collections::VecDeque;
+use std::path::PathBuf;
+use std::thread;
+
+use tokio::prelude::Future;
+
+use ipc_channel::ipc::{self, IpcReceiver, IpcSender};
+use net_traits::indexeddb_thread::{IndexedDBThreadMsg, IndexedDBThreadReturnType, SyncOperation};
+use servo_config::pref;
+use servo_url::ServoUrl;
+
+pub trait IndexedDBThreadFactory {
+    fn new(config_dir: Option<PathBuf>) -> Self;
+}
+
+impl IndexedDBThreadFactory for IpcSender<IndexedDBThreadMsg> {
+    fn new(config_dir: Option<PathBuf>) -> IpcSender<IndexedDBThreadMsg> {
+        let (chan, port) = ipc::channel().unwrap();
+
+        let mut idb_dir = PathBuf::new();
+        config_dir.map(|p| idb_dir.push(p));
+        idb_dir.push("IndexedDB");
+
+        thread::Builder::new()
+            .name("IndexedDBManager".to_owned())
+            .spawn(move || {
+                IndexedDBManager::new(port, RkvEngine::new(idb_dir.as_path())).start();
+            })
+            .expect("Thread spawning failed");
+
+        chan
+    }
+}
+
+struct IndexedDBManager<E: KvsEngine> {
```

Comment:
At a very high level, the spec reads that:

> Each origin has an associated set of databases. A database has zero or more object stores which hold the data stored in the database. http://w3c.github.io/IndexedDB/#ref-for-database

So I'm not sure how this would be best organized in terms of threading and underlying engines, and I do think you want a map somewhere of `ImmutableOrigin` -> `Database`. 

In script, to get the origin from which you're calling you can usually do `self.global().origin().immutable()`, which you could send along each message so that the manager appropriately route the request to the right database.

Reply:
Im working on a fix for this, this was done quite stupidly, and think I have figured out a really nice way to restructure this to allow multiple databases. The stores don't have to be namespaced once I am done with it either, so it will solve quite a bunch of issues